connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala

```diff
@@ -91,6 +93,37 @@ class SparkConnectPlanner(plan: proto.Relation, session: SparkSession) {
       transformRelation(rel.getInput))
   }
 
+  private def transformDeduplicate(rel: proto.Deduplicate): LogicalPlan = {
+    if (!rel.hasInput) {
+      throw InvalidPlanInput("Deduplicate needs a plan input")
+    }
+    if (rel.getAllColumnsAsKeys && rel.getColumnNamesCount > 0) {
+      throw InvalidPlanInput("Cannot deduplicate on both all columns and a subset of columns")
+    }
+    if (!rel.getAllColumnsAsKeys && rel.getColumnNamesCount == 0) {
```

Comment:
Why do we need `getAllColumnsAsKeys`? Seems like we can just tell when the columns are not set.

Reply:
and this is not matched with the logical plan.

Reply:
The issue is, in Spark connect client, we only see column names, not expr IDs. If the DF has duplicated column names, then deduplicate by all columns can't work in Spark connect client.

Reply:
actually, column name is unknown either, as the input plan is unresolved.

Reply:
I mean, we don't need `rel.getAllColumnsAsKeys` condition because we can know that's the case when `rel.getColumnNamesCount == 0`.

Reply:
ohh, okay. when it has all duplicated columns

Reply:
> I mean, we don't need `rel.getAllColumnsAsKeys` condition because we can know that's the case when `rel.getColumnNamesCount == 0`.

I want to clarify this specifically:

This is one of the Connect proto API design principle: we need to differentiate if a field is not set explicitly, or put it in another way, every intention should be expressed explicitly.  Ultimately, this is to avoid ambiguity on the API surface.

One example is Project. If we see a Project without anything in the project list, then how do we interpret that? Does the user want to indicate a `SELECT *`? Does the user actually generate an invalid plan. The problem now is there are two possibilities for a plan, and the worse part is, one possibility is a valid plan, another is not. This led us explicitly encode `SELECT *` into the proto https://github.com/apache/spark/pull/38023.

So one of the reasons that we have a bool flag here is to not use `rel.getColumnNamesCount == 0` to infer distinct on all columns which is to ambiguity.